### ISSUE TITLE
.Net 5.0 Application using NAudio libraries crashes without any error message

### ISSUE THREAD

**Issue by TheRealLenon** [dotnet/runtime#54132]
### Description
Hey Guys,

It seems something on .NET 5 made [NAudio (master-branch)](https://github.com/naudio/NAudio) crashing without any error message.

There is also an open [Issue](https://github.com/naudio/NAudio/issues/789) open on the NAudio´s GitHub.

After discussions i found out, that the runtime tries to access memory it does not own so Windows kills the process with Access Violation.

### Configuration
It happens on Windows 20H2 with **ANY** .NET 5.X.X version.

### Related information
Based on the [issue from NAudio](https://github.com/naudio/NAudio/issues/789), it was referred, that .NET Core 3.1 is working. I only could verify it for .NET 5.

**Comment by jackpoz** [comment]
Last week I installed Windows Updates and upgraded to Windows 10 21H1 after. I started a .NET 5 application that uses NAudio that worked fine the day before and after the upgrade it closed with Access Violation unhandled exception. Changing the runtime to .Net Core 3.1 fixed the issue.

**Comment by TheRealLenon** [comment]
Sadly I´m not being able to downgrade to .NET Core 3.1.

But referring to [this](https://github.com/naudio/NAudio/issues/789#issuecomment-859074097) comment, it is indicating, that there is a bug in the "hotfix" of the current .NET 5.x runtime, which lets the windows kernel kill the application with "Access Violation".

**Comment by Zintom** [comment]
It would be beneficial if anyone on Windows 20H2 could test this in .NET 5+, that way it would determine if a windows update was involved in the regression.

**Comment by TheRealLenon** [comment]
> It would be beneficial if anyone on Windows 20H2 could test this in .NET 5+, that way it would determine if a windows update was involved in the regression.

I´m currently running 20H2 and couldn´t run it with all .NET 5.x versions.

**Comment by jackpoz** [comment]
I tried on another computer with .NET 5.0.6 and Windows 10 20H2 and it works.
I downloaded .NET 5.0.7 runtime from https://dotnet.microsoft.com/download/dotnet/thank-you/runtime-desktop-5.0.7-windows-x64-installer , started the application and at first I heard random noises and then the application crashed

`dotnet --list-runtimes` before and working:
![image](https://user-images.githubusercontent.com/1153754/121880234-10cc6900-cd0e-11eb-8d0a-cbd5c1b23ba3.png)

`dotnet --list-runtimes` after and crashing:
![image](https://user-images.githubusercontent.com/1153754/121880279-1924a400-cd0e-11eb-840a-a5396b1e22b8.png)

So to me this is clearly a 5.0.7 issue not present in 5.0.6

**Comment by jkotas** [comment]
@jkoritzinsky Can this be regression introduced by #50882 ?

**Comment by wzchua** [comment]
It works if you published self-contained

**Comment by jackpoz** [comment]
I created a repro project at https://github.com/jackpoz/NetCore507Crash

![image](https://user-images.githubusercontent.com/1153754/121892245-c999a480-cd1c-11eb-95cb-70c3c421df95.png)

**Comment by jkotas** [comment]
https://github.com/jackpoz/NetCore507Crash fails for me with:
```
C:\NetCore507Crash>C:\NetCore507Crash\bin\Debug\net5.0\NetCore507Crash.exe
Playing some music...
Unhandled exception. System.InvalidOperationException: Buffer full
   at NAudio.Wave.BufferedWaveProvider.AddSamples(Byte[] buffer, Int32 offset, Int32 count)
   at NetCore507Crash.MusicPlayer.Play(String filePath) in C:\NetCore507Crash\Program.cs:line 51
   at NetCore507Crash.Program.Main() in C:\NetCore507Crash\Program.cs:line 20
```

**Comment by KyodaiKen** [comment]
To me it really looks like a memory pointer issue. It also is NOT the latest cumulative update. It's one of the .NET security updates. I've uninstalled the 2 latest cumulative updates and it didn't fix it. :/

**Comment by jackpoz** [comment]
@jkotas I added now a check in case the buffer gets full, I never reached that point in my tests with 5.0.7 as it crashes right away. Please pull latest version.

**Comment by jkoritzinsky** [comment]
I don't think this is caused by #50882 since that aligned all of the behavior to match .NET Core 3.1, but I guess it's possible that NAudio hits some corner case in it's interop. Without knowing if it even uses LayoutClass marshalling though, I don't know.

**Comment by jkotas** [comment]
It does use `LayoutClass` marshaling and it depends on blittable classes to be marshalled using pinning strategy. The problem is that 5.0.7 switched to use copying strategy for blittable classes that broke this library. 

The affected code is here: https://github.com/naudio/NAudio/blob/fb35ce8367f30b8bc5ea84e7d2529e172cf4c381/NAudio.WinMM/WaveOutBuffer.cs#L140 . This code depends on WaveHeader to be marshaled using pinning strategy.

Here is small test to demonstrate the problem:

On 5.0.6 (and .NET Framework), the program below prints pointers that are really far apart (GCHeap and stack):
On 5.0.7, the program below prints pointers that are really close togetger (two stack pointers):

```
using System;
using System.Runtime.InteropServices;
using System.IO;

[StructLayout(LayoutKind.Sequential)]
class WaveHeader
{
    /// <summary>pointer to locked data buffer (lpData)</summary>
    public IntPtr dataBuffer;
    /// <summary>length of data buffer (dwBufferLength)</summary>
    public int bufferLength;
    /// <summary>used for input only (dwBytesRecorded)</summary>
    public int bytesRecorded;
    /// <summary>for client's use (dwUser)</summary>
    public IntPtr userData;
    /// <summary>assorted flags (dwFlags)</summary>
    // public WaveHeaderFlags flags;
    /// <summary>loop control counter (dwLoops)</summary>
    public int loops;
    /// <summary>PWaveHdr, reserved for driver (lpNext)</summary>
    public IntPtr next;
    /// <summary>reserved for driver</summary>
    public IntPtr reserved;
}

unsafe class Program
{
    [DllImport("msvcrt.dll")]
    extern static void printf(string s, WaveHeader p);

    [DllImport("msvcrt.dll")]
    extern static void printf(string s, void* p);

    static unsafe void Main(string[] args)
    {
        int x;
        printf("%p\n", &x);
        printf("%p\n", new WaveHeader());
    }
}
```

@jkoritzinsky Could you please take it from here?

@jackpoz Thank you for providing good repro!

**Comment by jkoritzinsky** [comment]
I'm trying to figure out how this depends on the pinning optimization. I specifically added a compat feature as part of the change in 5.0.7 to automatically marshal both to and from native whenever a layoutclass with blittable contents is passed to ensure compat as well as tests to validate the behavior, so any dependency would have to be on the address itself and not the implicit side effects.

**Comment by jkotas** [comment]
> I'm trying to figure out how this depends on the pinning optimization. 

The `WaveHeader` address has to be valid even after `waveOutWrite` returns. The code achieves that by creating Pinned GCHandle for `WaveHeader`. This only works when the marshalled `WaveHeader` pointer points into the pinned object. It does not work when the marshalled `WaveHeader` pointer points to a stack local.

**Comment by jkoritzinsky** [comment]
So it does actually depend on the address. 😢

I'll put together a fix for this and once it's in we can decide if we want to backport it.

**Comment by michaelgsharp** [comment]
Whats interesting though is that it works on .NET 6 preview 3. So either the pinning optimization has been changed in preview 3 or something else is going on.

**Comment by michaelgsharp** [comment]
I am wondering it its a combination windows updates and .NET 5. Unless I am switching runtimes incorrectly (which is very possible) it doesn't work on any version of .NET 5 on my computer.

**Comment by KyodaiKen** [comment]
> 
> 
> I am wondering it its a combination windows updates and .NET 5. Unless I am switching runtimes incorrectly (which is very possible) it doesn't work on any version of .NET 5 on my computer.

Same here. 5.0 completely broken

**Comment by TheRealLenon** [comment]
> I am wondering it its a combination windows updates and .NET 5. Unless I am switching runtimes incorrectly (which is very possible) it doesn't work on any version of .NET 5 on my computer.

This is the same for me. Confusing, that this is not happening to all, as a few on here and on the NAudio Issue were able to run it with another .NET version.

**Comment by michaelgsharp** [comment]
So I tested on another machine that _only_ had 5.0.1 installed. It ran just fine. Tried same thing/version on the original machine same problem with crashing. Then, I uninstalled the June 2008 security update for 5.0.7 and now it works on my original machine. So it is the issue with 5.0.7 as mentioned above, what is still weird to me is why having 5.0.7 installed breaks all versions of .NET 5 even if you run with an older version.

**Comment by jkotas** [comment]
> what is still weird to me is why having 5.0.7 installed breaks all versions of .NET 5 even if you run with an older version.

Applications targeting 5.0.x will automatically roll forward to the latest 5.0.x installed by default.

**Comment by jkotas** [comment]
https://docs.microsoft.com/en-us/dotnet/core/versions/selection#framework-dependent-apps-roll-forward

**Comment by michaelgsharp** [comment]
I created a `global.json` file and set `"rollForward": "disable"`. Does that not prevent that behavior? I also tried specifically setting the SDK version in that file. And when using `dotnet run` I also tried manually setting the framework with `--fx-version`.

**Comment by jkoritzinsky** [comment]
I put out a PR for a fix in .NET 6. There's also a possible workaround. If NAudio were to make their interop classes `sealed`, then this would also be fixed on 5.0.7.

**Comment by michaelgsharp** [comment]
@jkoritzinsky in .NET 6 preview 3 this runs fine without issues already.

**Comment by jkoritzinsky** [comment]
The same fix that went into .NET 5 that broke this also went into .NET 6 in the Preview 4 time frame, so a fix is needed for .NET 6 as well.

**Comment by Fabi** [comment]
> The same fix that went into .NET 5 that broke this also went into .NET 6 in the Preview 4 time frame, so a fix is needed for .NET 6 as well.

fun fact: it works fine on 5.0.8 and 6 preview 4 and later for me

**Comment by AraHaan** [comment]
Perhaps some other code in .NET 6 Preview 4 that made it work was forgotten in the backport and it just works in .NET 6 Preview 4 because of coincidence.

**Comment by jkotas** [comment]
The crash is caused by use-after-free bug. The crash will only happen if something else actually modified the freed memory in a specific window of time. It is certainly possible that the crash will come and go due to unrelated changes - that's the nature of use-after-free bugs.

**Comment by jkoritzinsky** [comment]
Would anyone here be willing/able to test that their NAudio scenario is fixed if I share a .NET runtime build with the proposed fix?

**Comment by cits** [comment]
> 
> 
> Would anyone here be willing/able to test that their NAudio scenario is fixed if I share a .NET runtime build with the proposed fix?

I am willing test my NAudio WPF app.

**Comment by jcddcjjcd** [comment]
I am also ready to try with my WPF app.

**Comment by TheRealLenon** [comment]
> Would anyone here be willing/able to test that their NAudio scenario is fixed if I share a .NET runtime build with the proposed fix?

I would test it with input and instant output on console applications.

**Comment by warappa** [comment]
Maybe worth noting, that I'm running NAudio 2.0.0 on 5.0.***6*** as I have _not_ 5.0.***7*** installed and having a very similar, but slightly different problem: `AudioDG.exe` crashes in background while the WPF app keeps on running.

### Maybe Helpful Details
In WPF I use NAudio to only change volume on default audio device and play a test-sound with `System.Media.SystemSounds.Beep.Play()` (read: not played by NAudio). After hearing the beep-sound in *full length*, `AudioDG.exe` crashes. The sound device gets disconnected by Windows(?) and 1 second later added again (I can even see it vanish and reappear in `Sound` settings of classic control panel).

### Question
Would this still be explicable with the found bug, or is this then maybe another issue?

### Event viewer (multiple entries like this)
```
Name der fehlerhaften Anwendung: AUDIODG.EXE, Version: 10.0.19041.1023, Zeitstempel: 0x4507cb5a
Name des fehlerhaften Moduls: ntdll.dll, Version: 10.0.19041.1023, Zeitstempel: 0x7977b9de
Ausnahmecode: 0xc0000005
Fehleroffset: 0x0000000000063416
ID des fehlerhaften Prozesses: 0x904
Startzeit der fehlerhaften Anwendung: 0x01d762b2305f2407
Pfad der fehlerhaften Anwendung: C:\Windows\system32\AUDIODG.EXE
Pfad des fehlerhaften Moduls: C:\Windows\SYSTEM32\ntdll.dll
Berichtskennung: 70721d07-491f-4dcb-a91d-b69e9da2cdd8
Vollständiger Name des fehlerhaften Pakets: 
Anwendungs-ID, die relativ zum fehlerhaften Paket ist: 
```

### `dotnet --info` in WPF app folder
```
.NET SDK (gemäß "global.json"):
 Version:   5.0.300
 Commit:    2e0c8c940e

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.300\

Host (useful for support):
  Version: 5.0.6
  Commit:  478b2f8c0e

.NET SDKs installed:
  2.1.807 [C:\Program Files\dotnet\sdk]
  2.2.207 [C:\Program Files\dotnet\sdk]
  5.0.202 [C:\Program Files\dotnet\sdk]
  5.0.300-preview.21258.4 [C:\Program Files\dotnet\sdk]
  5.0.300 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by jkotas** [comment]
> Would this still be explicable with the found bug, or is this then maybe another issue?

It is very unlikely to be related to this bug. I would recommend reporting it via Windows Feedback. It looks like a Windows OS bug to me.

**Comment by jcddcjjcd** [comment]
This may also not be related but I mention it anyway.

I have stopped using NAudio to play wav files but still use the Volume get and set in NAudio and GetDefaultRenderDevice().AudioMeterInformation.MasterPeakValue.
I have a very rare bug now which I suspect may come from NAudio PeakValues. It does not crash the app but does destroy the RealTek audio driver requiring a reinstall. The crash report mentions ADAudio from Realtek and the NAudio PeakValues code.

This has not happened prior to net v5.07

**Comment by jkoritzinsky** [comment]
For everyone who wants to validate this, I've added a link to a zip at the end of this comment with an x64 runtime. If you set the `DOTNET_ROOT` environment variable to the folder where you unzip this file and set the `DOTNET_MULTILEVEL_LOOKUP` environment variable to `0` before you run your application from the app host (the .exe that matches your project name), your application will run against the local copy.

https://1drv.ms/u/s!Aj0M_O5AaaC_leEPU6s6nPuz_VGrdA?e=sRRZ7s

**Comment by cits** [comment]
> For everyone who wants to validate this, I've added a link to a zip at the end of this comment with an x64 runtime. If you set the `DOTNET_ROOT` environment variable to the folder where you unzip this file and set the `DOTNET_MULTILEVEL_LOOKUP` environment variable to `0` before you run your application from the app host (the .exe that matches your project name), your application will run against the local copy.
> 
> https://1drv.ms/u/s!Aj0M_O5AaaC_leEPU6s6nPuz_VGrdA?e=sRRZ7s

Thanks. Worked with my NAudio WPF app.

Any word on how long this fix will make it into an official release? Days, weeks or months?

**Comment by jkoritzinsky** [comment]
Reopening to track the servicing fix.

**Comment by jkoritzinsky** [comment]
This will likely be more on the scale of months rather than weeks due to the servicing timeline.

**Comment by cits** [comment]
> 
> 
> This will likely be more on the scale of months rather than weeks due to the servicing timeline.

OK. That's disappointing,, I understand.
Time to implement my plan "B", drop using NAudio.

**Comment by jcddcjjcd** [comment]
The sad thing here is that NAudio is widely used. It has 2.34 Million downloads on github and there is currently nothing that comes close to it's functionality that is maintained.
There was mention above that if NAudio made it's classes sealled then the problem would go away.
Is that an option that could be considered?

**Comment by jkoritzinsky** [comment]
That option would have to be considered by the owners of NAudio, and based off the parallel issue over there, it seems it might be.

**Comment by markheath** [comment]
A new version of NAudio is available (2.0.1) that resolves this issue by making `WaveHeader` sealed.

**Comment by jkoritzinsky** [comment]
Fixed by #54244
